core/shared/src/main/scala/zio/internal/metrics/ConcurrentMetricRegistry.scala

```diff
@@ -18,32 +18,55 @@ private[zio] class ConcurrentMetricRegistry {
 
   def snapshot()(implicit unsafe: Unsafe): Set[MetricPair.Untyped] = {
     val iterator = map.entrySet().iterator()
-    val result   = scala.collection.mutable.Set[MetricPair.Untyped]()
+    val result   = Set.newBuilder[MetricPair.Untyped]
+    result.sizeHint(map.size())
     while (iterator.hasNext) {
       val value = iterator.next()
-      val key   = value.getKey()
-      val hook  = value.getValue()
-      result.add(MetricPair.make(key, hook.get()))
+      val key   = value.getKey
+      val hook  = value.getValue
+      result += MetricPair.make(key, hook.get())
     }
-    result.toSet
+    result.result()
   }
 
   def get[Type <: MetricKeyType](
     key: MetricKey[Type]
   )(implicit unsafe: Unsafe): MetricHook[key.keyType.In, key.keyType.Out] = {
     type Result = MetricHook[key.keyType.In, key.keyType.Out]
 
-    var hook0: MetricHook[_, zio.metrics.MetricState.Untyped] = map.get(key)
-
-    if (hook0 eq null) {
-      (key.keyType match {
-        case MetricKeyType.Counter             => getCounter(key.asInstanceOf[MetricKey.Counter])
```

Comment:
In this case, we'd make 3 `.get` calls:
1. In the current method: `var hook0: MetricHook[_, zio.metrics.MetricState.Untyped] = map.get(key)`
2. In `getCounter`: `var value = map.get(key)`
3. In `getCounter`: `value = map.get(key)`

I wonder why it was implemented like this 🤔

Reply:
Both the current and proposed implementations are trade offs between availability and correctness: `computeIfAbsent` do computation of absent value in a synchronized block (so less availability at highly concurrent access, that could be sources of unwanted latency spikes) but it avoids races when some metrics overwrite themself for the same key (so it avoids wrong metric values). I would prefer correctness with `computeIfAbsent` and for reducing of latency spikes I would pre-initialize all metrics before serving concurrent requests.

Reply:
The `computeIfAbsent` does computation in a synchronized block. That could be source of unwanted latency spikes during highly concurrent access. But from other side the current solution could lead to unwanted spikes of metric hook creations that could be also costly in terms of CPU usage and memory bandwidth saturation at highly concurrent access.